sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotDecreaseMemberVisibility.cs

```diff
@@ -0,0 +1,198 @@
+                    {
+                        yield return member;
+                    }
+
+                    classType = classType.BaseType;
+                }
+            }
+
+            public Diagnostic FindIssue(MemberDeclarationSyntax memberDeclaration)
+            {
+                var memberSymbol = semanticModel.GetDeclaredSymbol(memberDeclaration);
+
+                var methodSymbol = memberSymbol as IMethodSymbol;
+                if (methodSymbol != null)
+                {
+                    var hidingMethod = allBaseClassMethods.FirstOrDefault(
+                        m => DecrasesAccess(m.DeclaredAccessibility, methodSymbol.DeclaredAccessibility) &&
+                             IsMatchingSignature(m, methodSymbol));
+
+                    if (hidingMethod != null)
+                    {
+                        var location = (memberDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation();
+                        if (location != null)
+                        {
+                            return Diagnostic.Create(rule, location, hidingMethod);
+                        }
+                    }
+
+                    return null;
+                }
+
+                var propertySymbol = memberSymbol as IPropertySymbol;
+                if (propertySymbol != null)
+                {
+                    var hidingProperty = allBaseClassProperties.FirstOrDefault(p => DecreasesPropertyAccess(p, propertySymbol));
+                    if (hidingProperty != null)
+                    {
+                        var location = (memberDeclaration as PropertyDeclarationSyntax).Identifier.GetLocation();
+                        return Diagnostic.Create(rule, location, hidingProperty);
+                    }
+                }
+                return null;
+            }
+
+            private static bool DecreasesPropertyAccess(IPropertySymbol baseProperty, IPropertySymbol propertySymbol)
+            {
+                if (baseProperty.Name != propertySymbol.Name ||
+                    !Equals(baseProperty.Type, propertySymbol.Type))
+                {
+                    return false;
+                }
+
+                var baseGetAccess = GetEffectiveDeclaredAccess(baseProperty.GetMethod, baseProperty.DeclaredAccessibility);
+                var baseSetAccess = GetEffectiveDeclaredAccess(baseProperty.SetMethod, baseProperty.DeclaredAccessibility);
+
+                var propertyGetAccess = GetEffectiveDeclaredAccess(propertySymbol.GetMethod, baseProperty.DeclaredAccessibility);
+                var propertySetAccess = GetEffectiveDeclaredAccess(propertySymbol.SetMethod, baseProperty.DeclaredAccessibility);
+
+                return DecrasesAccess(baseGetAccess, propertyGetAccess) ||
+                       DecrasesAccess(baseSetAccess, propertySetAccess);
+            }
+
+            private static Accessibility GetEffectiveDeclaredAccess(IMethodSymbol method, Accessibility propertyDefaultAccess)
```

Comment:
Maybe this is not doing what you want but the class `SymbolHelper` class already has a `GetEffectiveAccessiblity` method.